### PR TITLE
TM: Update and Refactor

### PR DESCRIPTION
**Pull request by marcbonnici**
TM has been updated and restructured to be clearer, including splitting off
assistant and creating a runtime parameter manager in their own files.

### CODE REVIEW

File: wa/framework/target/runtime_config.py [hunk 1]
```diff
@@ -26,67 +38,136 @@ def core_names(self):
...
+                                     constraint=lambda x, name=name:
+                                             0 < x <= len(self.target.core_cpus(name)),
+                                     description="""
+                                     The number of {} cores to be hotplugged
+                                     """.format(name),
+                                     setter=self.set_num_cores,
+                                     setter_params={'core': name})
+
+        for cpu_no in range(self.target.number_of_cpus):
+            param_name = 'num_cpu{}_cores'.format(cpu_no)
```
Comment (setrofim) [inline wa/framework/target/runtime_config.py]: "num_cpu#_cores"? Maybe rename to "cpu#_online" and change to boolean.

File: wa/framework/target/runtime_config.py [hunk 2]
```diff
@@ -26,67 +38,136 @@ def core_names(self):
...
+        '''Online all CPUs required before then off-lining'''
+        for c, n in self.num_cores.iteritems():
             self.set_num_online_cpus(c, n)
+        for c, n in self.num_cores.iteritems():
+            self.set_num_offline_cpus(c, n)
 
     def clear(self):
         self.num_cores = defaultdict(dict)
 
     def set_num_online_cpus(self, core, number):
```
Comment (setrofim) [inline wa/framework/target/runtime_config.py]: should be rnamed to "online_cpus()", as this doesn't set the number of online cpus, but onlines specific cpus.

File: wa/framework/target/runtime_config.py [hunk 3]
```diff
@@ -26,67 +38,136 @@ def core_names(self):
...
-        indexes = [i for i, c in enumerate(self.target.core_names) if c == core]
-        self.target.hotplug.online(*indexes[:number])
-        self.target.hotplug.offline(*indexes[number:])
+        cpus = resolve_cpus(core, self.target)
+        self.target.hotplug.online(*sorted(cpus[:number]))
+
+    def set_num_offline_cpus(self, core, number):
```
Comment (setrofim) [inline wa/framework/target/runtime_config.py]: should be renamed to `offline_cpus`, as this doesn't set the number of offline cpus but offlines specific cpus.

File: wa/framework/target/runtime_config.py [hunk 4]
```diff
@@ -96,28 +177,44 @@ class SysfileValuesRuntimeConfig(RuntimeConfig):
...
             if f.endswith('+'):
                 f = f[:-1]
             elif f.endswith('+!'):
                 f = f[:-2] + '!'
             else:
                 if f.endswith('!'):
-                    self._check_exists(f[:-1])
+                    obj.check_exists(f[:-1])
                 else:
-                    self._check_exists(f)
```
Comment (setrofim) [inline wa/framework/target/runtime_config.py]: I'm not sure I follow the above logic. What do `'+'` and `'!'` represent?

File: wa/framework/target/runtime_config.py [hunk 5]
```diff
@@ -1,23 +1,35 @@
...
+
+
+class RuntimeParameter(Parameter):
+    def __init__(self, name, setter, setter_params, **kwargs):
+        super(RuntimeParameter, self).__init__(name, **kwargs)
+        self.setter = setter
+        self.setter_params = setter_params
+
+    def set(self, obj, value):
+        self.setter(obj, value, **self.setter_params)
```
Comment (setrofim) [inline wa/framework/target/runtime_config.py]: should call `self.validate_value(self.name, value)` before invoking the setter

Comment (marcbonnici) [inline wa/framework/target/runtime_config.py]: Updated.

**Pull request merged**